bpf: ICMP checksum handling in SNAT RevNAT

yushoyamaguchi opened the pull request:
Fixes: https://github.com/cilium/cilium/issues/40827

When a LB node forwards packets to a backend-pod on another node,
the packets are SNATed.
The corresponding reverse NAT (RevNAT) is then applied
to the reply packets.

Checksum recaluculation is required when RevNAT.
When the reply packet is an ICMP error packet, the checksum
in the ICMP header must be recalculated to reflect the changes
of the inner packet header.
https://datatracker.ietf.org/doc/html/rfc5508#section-4.1

Since that logic was missing, it has been added in this commit.


The below is an error message of `nat4_icmp_error_tcp_snat_revnat` ([new test](https://github.com/cilium/cilium/pull/43196/commits/c4fff34f4e484af5328add58afff0f007a0ad002)) without the [checksum recalculate patch](https://github.com/cilium/cilium/pull/43196/commits/8495c0c5c063e2bd8c896e74f7d636f69666d18f).
<details><summary> <strong>An error message of RevNAT BPF-test</strong> </summary>

```
--- FAIL: TestBPF (39.66s)
--- FAIL: TestBPF/icmp_error_revnat.o (0.55s)
--- FAIL: TestBPF/icmp_error_revnat.o/nat4_icmp_error_tcp_snat_revnat (0.42s)

    bpf_test.go:577: icmp_error_revnat.c:137: CTX and buffer 'ICMP4_ERR_FRAG_NEEDED_AFTER_REVNAT' content mismatch 
    bpf_test.go:421:   Scapy asserts failed: 1
    bpf_test.go:428: 
        === START ./scapy.h:171 'icmp4_revnat_ok' ===
        >>> Expected (len: 164 bytes) <<<
        ###[ Ethernet ]### 
          dst       = 13:37:13:37:13:37
          src       = de:ad:be:ef:de:ef
          type      = IPv4
        ###[ IP ]### 
             version   = 4
             ihl       = 5
             tos       = 0x0
             len       = 68
             id        = 1
             flags     = 
             frag      = 0
             ttl       = 64
             proto     = icmp
             chksum    = 0xa60d
             src       = 192.168.0.2
             dst       = 10.0.10.1
             \options   \
        ###[ ICMP ]### 
                type      = dest-unreach
                code      = fragmentation-needed
                chksum    = 0xcbe5
                reserved  = 0
                length    = 0
                nexthopmtu= 1500
                unused    = ''
        ###[ IP in ICMP ]### 
                   version   = 4
                   ihl       = 5
                   tos       = 0x0
                   len       = 40
                   id        = 1
                   flags     = DF
                   frag      = 0
                   ttl       = 64
                   proto     = tcp
                   chksum    = 0x6624
                   src       = 10.0.10.1
                   dst       = 192.168.0.2
                   \options   \
        ###[ TCP in ICMP ]### 
                      sport     = 3030
                      dport     = http
                      seq       = 0
                      ack       = 0
                      dataofs   = 5
                      reserved  = 0
                      flags     = S
                      window    = 8192
                      chksum    = 0xaf11
                      urgptr    = 0
                      options   = ''
        
        >>> Got (len: 164 bytes) <<<
        ###[ Ethernet ]### 
          dst       = 13:37:13:37:13:37
          src       = de:ad:be:ef:de:ef
          type      = IPv4
        ###[ IP ]### 
             version   = 4
             ihl       = 5
             tos       = 0x0
             len       = 68
             id        = 1
             flags     = 
             frag      = 0
             ttl       = 64
             proto     = icmp
             chksum    = 0xa60d
             src       = 192.168.0.2
             dst       = 10.0.10.1
             \options   \
        ###[ ICMP ]### 
                type      = dest-unreach
                code      = fragmentation-needed
                chksum    = 0x788e
                reserved  = 0
                length    = 0
                nexthopmtu= 1500
                unused    = ''
        ###[ IP in ICMP ]### 
                   version   = 4
                   ihl       = 5
                   tos       = 0x0
                   len       = 40
                   id        = 1
                   flags     = DF
                   frag      = 0
                   ttl       = 64
                   proto     = tcp
                   chksum    = 0x6624
                   src       = 10.0.10.1
                   dst       = 192.168.0.2
                   \options   \
        ###[ TCP in ICMP ]### 
                      sport     = 3030
                      dport     = http
                      seq       = 0
                      ack       = 0
                      dataofs   = 5
                      reserved  = 0
                      flags     = S
                      window    = 8192
                      chksum    = 0xaf11
                      urgptr    = 0
                      options   = ''
        
        >>> Diff <<<
          --- a/pkt (Expected)
          +++ b/pkt (Got)
        
          - [ Ether ][ IP ][ ICMP ].chksum: 52197
          + [ Ether ][ IP ][ ICMP ].chksum: 30862
        
        === END ./scapy.h:171 'icmp4_revnat_ok' ===

┌──────────────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │                  PACKAGE                   │ COVER │ PASS │ FAIL │ SKIP  │
│─────────┼─────────┼────────────────────────────────────────────┼───────┼──────┼──────┼───────│
│  FAIL   │ 39.70s  │ github.com/cilium/cilium/bpf/tests/bpftest │  --   │ 594  │  3   │  0    │
└──────────────────────────────────────────────────────────────────────────────────────────────┘
```

</details>


```release-note
Fix ICMP error packet handling by adding the missing checksum recalculation performed during RevNAT for SNATed load-balanced traffic.
```

yushoyamaguchi committed:
ICMP csum handling in SNAT RevNAT

When a LB node forwards packets to a backend-pod on another node,
the packets are SNATed.
The corresponding reverse NAT (RevNAT) is then applied
to the reply packets.

Checksum recaluculation is required when RevNAT.
When the reply packet is an ICMP error packet, the checksum
in the ICMP header must be recalculated to reflect the changes
of the inner packet header.
https://datatracker.ietf.org/doc/html/rfc5508#section-4.1

Since that logic was missing, it has been added in this commit.

Fixes: https://github.com/cilium/cilium/issues/40827

Signed-off-by: Yusho Yamaguchi <ysh.824@outlook.jp>

pchaigno commented on bpf/lib/nat.h:
```diff
@@ -1074,10 +1126,25 @@ snat_v4_rev_nat_handle_icmp_error(struct __ctx_buff *ctx,
+						 tuple.dport, (*state)->to_dport, port_off,
+						 outer_icmp_off,
+						 icmp_has_inner_l4_csum &&
+						 is_inner_l4_csum_enabled);
```
I'm unsure what our existing BPF tests (`bpf/tests/`) look for this, do we already have anything? Could we extend it to cover this change?

yushoyamaguchi replied:
Sorry, I didn’t think about that.
I'll check if existing tests can testify my logic.

yushoyamaguchi replied:
Do we need to verify the checksum in this test?
Currently, packets are being sent with an incorrect checksum, but all existing BPF tests are still passing.

Personally, I think that building a framework for checksum verification inside the BPF tests could be handled separately. Introducing checksum validation into the BPF tests would likely cause some of the existing code to start failing.

julianwiedmann replied:
@msune recently added scapy support for BPF tests (https://github.com/cilium/cilium/pull/40294), I think that would be a perfect fit for validating these fixes. Could you please see if that's doable with a reasonable effort?

yushoyamaguchi replied:
Please let me check whether my understanding of checksum verification in BPF tests that use scapy is correct.

By using `ASSERT_CTX_BUF_OFF`, we can compare the checksum field of a packet produced by Cilium’s BPF functions with the checksum field of the packet generated on the scapy side, right?

If that’s the case, then simply switching the SNAT-RevNAT tests to use scapy-generated packets should already allow us to automatically verify the checksum as well, right?

msune replied:
`ASSERT_CTX_BUF_OFF()` compares the CTX contents with a scapy buffer, see:

https://github.com/cilium/cilium/blob/3ab1b9a54fa0e0e1b977025302ff9199f4af366d/bpf/tests/scapy.h#L234

It is meant to be used to compare entire packets, or the first part of it (`LEN`). Scapy - unless explicitely set - calculates and fills in L3 and L4 checksum.

So yes, using that assert will validate the checksum. Mind some unit tests store the return code in the first 4bytes of CTX.

I suggest you to read and follow this guide to port the test to scapy:

https://github.com/cilium/cilium/tree/main/bpf/tests/scapy#how-to-port-or-write-your-first-unit-test-with-scapy-buffers

yushoyamaguchi replied:
@msune cc @julianwiedmann 
Okay, thank you!

https://github.com/cilium/cilium/blob/2c59ba7dfaa5065e3b71814b97be7d217e07cda3/bpf/tests/bpf_nat_tests.c
There are already many tests cases for validating `snat_v4_rev_nat()`.
- Should I rewrite these existing test cases to use _scapy_?
- Or is it sufficient for me to simply add new a test case that use _scapy_?
<br>
Is there any motivation or long-term direction to gradually replace the existing test cases with scapy-based ones?

yushoyamaguchi replied:
@julianwiedmann   @msune 
I've implemented a new test case using scapy and I've confirmed this test case could also validate checksum.
Please review.

julianwiedmann replied:
Awesome, thank you! You could even quote the test's error output (without your fix) in the patch description.

yushoyamaguchi replied:
After I fix the test case to address your feedback, I'll put the error message into this patch!

yushoyamaguchi replied:
@julianwiedmann 
I've added! 
https://github.com/cilium/cilium/pull/43196#issue-3703588614

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -528,6 +528,56 @@ snat_v4_rewrite_headers(struct __ctx_buff *ctx, __u8 nexthdr, int l3_off,
+	}
+
+	return 0;
+}
```
(More as inspiration, this wouldn't necessarily need to happen as part of this PR)

Now that we have logic to calculate the `diff` that needs to be applied on the outer packet - can we hand this diff back to the caller, and have **them** apply it (as part of re-writing the outer headers) ? This would probably be an additional `csum_diff` parameter for `snat_v4_rewrite_headers()`.

yushoyamaguchi replied:
Please let me confirm what you said.
You said that `snat_v4_rewrite_icmp_error_headers()` should just calculate `diff` for outer-header and pass it to `snat_v4_rev_nat_handle_icmp_error()` , then `snat_v4_rev_nat_handle_icmp_error()` apply the `diff` to outer-header, right?

yushoyamaguchi replied:
How is `snat_v4_rewrite_headers()` related to your plan?
I guess  `snat_v4_rewrite_headers()` is for rewriting inner-header, not outer-header.

julianwiedmann replied:
> Please let me confirm what you said.
> You said that `snat_v4_rewrite_icmp_error_headers()` should just calculate `diff` for outer-header and pass it to `snat_v4_rev_nat_handle_icmp_error()` , then `snat_v4_rev_nat_handle_icmp_error()` apply the `diff` to outer-header, right?

The picture I have in my head is that `snat_v4_rev_nat_handle_icmp_error()` modifies the inner headers, and as part of doing so calculates the resulting `diff`. It then reports this `diff` back to `snat_v4_rev_nat()`, which forwards it to `snat_v4_rewrite_headers()`. There the `diff` is mixed into the rewrite of the outer L4 header's `csum` field.

This for instance would allow us to avoid the additional special-casing for UDP - the outer layer will already check for this, and skip the L4 csum update accordingly.

yushoyamaguchi replied:
@julianwiedmann 
I've understood. Thank you!
I've fixed.

yushoyamaguchi committed:
bpf-test of SNAT RevNAT using scapy

This commit adds a new test case of SNAT RevNAT using scapy.
By this test case, we could also validate checksum of outputed packet.

Signed-off-by: Yusho Yamaguchi <ysh.824@outlook.jp>

julianwiedmann commented on bpf/tests/icmp_error_revnat.c:
```diff
@@ -0,0 +1,140 @@
+/* IP addresses mapping to Scapy definitions (in host byte order):
+ * v4_node_one   = "10.0.10.1"  -> IP_HOST (the node being tested)
+ * v4_node_two   = "10.0.10.2"  -> IP_ROUTER (router sending ICMP error)
+ * v4_ext_one    = "110.0.11.1" -> IP_ENDPOINT (external endpoint)
```
```suggestion
 * v4_pod_one    = "110.0.11.1" -> IP_ENDPOINT
```

Since this is supposed to be a local endpoint that requires masquerading on egress, calling it "external" feels very confusing to me.

yushoyamaguchi replied:
@julianwiedmann 
I used inappropriate IP addresses for SNAT RevNAT in this test.
I've fixed.
Now, IP addresses of **outer** header are shown below.
- Before SNAT Revnat-> src: `pod_two`, dst: `pod_one`
- After SNAT Revnat(Expected)-> src: `pod_two`, dst: `node_one`

Is it appropriate?

yushoyamaguchi replied:
Actual packet in my kind cluster

Before RevNAT
<img width="1237" height="360" alt="スクリーンショット 2026-01-10 113650" src="https://github.com/user-attachments/assets/072b6c1b-6ba8-4620-b0ce-e48b6595f30f" />

After RevNAT
<img width="1220" height="295" alt="スクリーンショット 2026-01-10 113727" src="https://github.com/user-attachments/assets/18aec01b-b29d-4a84-9bf7-0ed785768e54" />

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -523,11 +524,46 @@ snat_v4_rewrite_headers(struct __ctx_buff *ctx, __u8 nexthdr, int l3_off,
 	return 0;
 }
 
+static __always_inline void
+snat_v4_calc_icmp_error_csum_diff(__be32 old_addr, __be32 new_addr,
+				  __be16 old_port, __be16 new_port,
+				  bool inner_has_l4_csum, __wsum *diff_for_csum)
+{
+	__be32 old_port32 = (__be32)old_port;
+	__be32 new_port32 = (__be32)new_port;
+
+	*diff_for_csum = 0;
+
+	if (inner_has_l4_csum) {
+		/* Calculate diff value for checksum.
+		 * Reflect the change in the inner L4 checksum caused by the pseudo-address update
+		 * into diff_for_csum.
+		 * All the other changes in inner packet cancel each other out.
+		 */
+		if (old_addr != new_addr)
+			*diff_for_csum = csum_diff(&new_addr, 4, &old_addr, 4, 0);
+	} else {
+		/* Calculate diff value for checksum.
+		 * If the inner L4 header does not include the L4 checksum,
+		 * only the port is modified within the inner L4 header.
+		 */
+		if (old_port != new_port)
+			*diff_for_csum = csum_diff(&old_port32, 4, &new_port32, 4, 0);
+	}
+}
+
```
I think eventually we'll want to integrate this logic into `snat_v4_rewrite_headers()` (so it literally tells the caller what csum-relevant changes it made, and the caller can then feed that information into the next stage of header rewrites). But not today.

:exploding_head: for the "truncated L4" scenario, this is such an annoying corner case to handle ...

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -523,11 +524,46 @@ snat_v4_rewrite_headers(struct __ctx_buff *ctx, __u8 nexthdr, int l3_off,
 		    csum_l4_replace(ctx, l4_off, &csum, 0, sum, flags) < 0)
 			return DROP_CSUM_L4;
+
+		/* Apply additional L4 checksum diff if provided (for ICMP error messages). */
+		if (l4_csum_diff_from_inner && !csum.offset) {
+			csum.offset = offsetof(struct icmphdr, checksum);
+			if (csum_l4_replace(ctx, l4_off, &csum, 0, l4_csum_diff_from_inner, 0) < 0)
+				return DROP_CSUM_L4;
+		}
```
For the future - I wonder whether we can integrate this with the `l4_modify_port()` path.

julianwiedmann reviewed (commented):
Thank you! Looks good to me overall - I added a few comments, although I think we shouldn't attempt to optimize the logic further for now. Let's do that once we have addressed the bug.

One aspect I'd still want to confirm is what the effect on `skb->csum` is. Not sure whether we can easily validate that from within the BPF test suite today.

msune commented on bpf/tests/scapy/icmp_err_pkt_defs.py:
```diff
@@ -0,0 +1,61 @@
+from pkt_defs_common import *
+
+icmp4_frag_needed = (
+    Ether(
```
We don't have a linter here, but can you please use the same format as in here for consistency:

https://github.com/yushoyamaguchi/cilium/blob/142715dab600d1be8c456dd345326f6811d4df11/bpf/tests/scapy/tc_l2_announce_pkt_defs.py#L9-L12

That is, generally one Layer per line with all fields there.

msune commented on bpf/tests/scapy/pkt_defs.py:
```diff
@@ -11,3 +11,4 @@
 from tc_l2_announce_pkt_defs import *
 from tc_l2_announce6_pkt_defs import *
 from wg_from_netdev_pkt_defs import *
+from icmp_err_pkt_defs import *
```
Very long but since this is specific to revnat only:

`icmp_err_revnat_pkt_defs.py`

msune commented on bpf/tests/icmp_error_revnat.c:
```diff
@@ -0,0 +1,140 @@
+	pktgen__init(&builder, ctx);
+
+	/* Use Scapy-generated ICMP error packet */
+	BUF_DECL(ICMP4_FRAG_NEEDED, icmp4_frag_needed);
```
`ICMP4_FRAG_NEEDED` is a rather generic name. Buffer name (today), are global. I'd suggest something more specific to this test.

msune commented on bpf/tests/icmp_error_revnat.c:
```diff
@@ -0,0 +1,140 @@
+	/* Compare the packet with expected output after rev-NAT.
+	 * Note: offset sizeof(__u32) to skip the return code prepended by framework.
+	 */
+	BUF_DECL(EXPECTED_PKT, icmp4_frag_needed_after_revnat);
```
Same thing here: `EXPECTED_PKT` is too generic. Prefix with the test name.

msune reviewed (changes requested):
I only reviewed the Scapy part. Overall LGTM. Minor changes to style + naming required.

Thanks!

julianwiedmann reviewed (approved):
Thank you, looks good to me!

So as follow-up actions, we'd want to
1. add more scapy tests (IPv4 and IPv6),
2. fix up the forward path (`snat_v4_nat_handle_icmp_error()`),
3. look at the suggested optimizations
